File: python/src/agents/clause_extraction.py
"""
条款提取Agent — 多Agent流水线第一步。

职责：
1. 解析合同文档 (PDF/DOCX → 纯文本)
2. 按条款切分文本
3. 识别条款类别 (付款、违约、保密等)
4. 提取关键实体 (甲乙方、金额、日期、标的物)
5. 输出结构化条款数据
"""

from __future__ import annotations

import json
import logging

from langchain_core.messages import HumanMessage, SystemMessage

from ..config import get_llm
from ..parsers.document_parser import DocumentParser, parse_raw_text_to_sections
from ..pipeline.state import (
    Clause,
    ClauseCategory,
    ContractEntity,
    ContractReviewState,
    ReviewStatus,
)

logger = logging.getLogger(__name__)

SYSTEM_PROMPT = """你是一个专业的法律合同条款提取专家。你的任务是从合同文本中提取结构化条款信息。

请严格按照以下JSON格式输出（不要输出其他内容）：

{
  "contract_type": "合同类型（如：买卖合同、劳动合同、租赁合同等）",
  "clauses": [
    {
      "title": "条款标题",
      "content": "条款完整内容",
      "category": "条款类别",
      "section_number": "章节编号"
    }
  ],
  "entities": [
    {
      "entity_type": "实体类型",
      "value": "实体值",
      "location": "出现位置"
    }
  ]
}

条款类别(category)只能是以下之一：
- payment: 付款条款
- liability: 责任条款
- confidentiality: 保密条款
- termination: 终止/解除条款
- intellectual_property: 知识产权条款
- dispute_resolution: 争议解决条款
- force_majeure: 不可抗力条款
- warranty: 保证/担保条款
- indemnification: 赔偿条款
- governing_law: 适用法律条款
- other: 其他条款

实体类型(entity_type)包括：
- party_a: 甲方
- party_b: 乙方
- amount: 金额
- date: 日期
- subject: 标的物/服务内容
- duration: 期限
- address: 地址"""


class ClauseExtractionAgent:
    """条款提取Agent：合同文档解析 + 条款分割 + 实体识别 + 条款分类。"""

    def __init__(self):
        self.parser = DocumentParser()
        self.llm = None

    def _ensure_llm(self):
        if self.llm is None:
            self.llm = get_llm()

    def __call__(self, state: ContractReviewState) -> dict:
        """LangGraph节点函数：执行条款提取。"""
        logger.info("条款提取Agent开始处理 review_id=%s", state.review_id)

        try:
            raw_text = state.raw_text
            if not raw_text and state.document_path:
                raw_text = self.parser.parse(state.document_path)

            if not raw_text:
                return {
                    "errors": state.errors + ["合同文本为空，无法进行条款提取"],
                    "status": ReviewStatus.COMPLETED,
                }

            sections = parse_raw_text_to_sections(raw_text)
            text_for_llm = self._prepare_text(raw_text, sections)

            self._ensure_llm()
            result = self._extract_with_llm(text_for_llm)

            clauses = self._parse_clauses(result)
            entities = self._parse_entities(result)
            contract_type = result.get("contract_type", "未识别")

            logger.info(
                "条款提取完成：%d个条款，%d个实体，合同类型=%s",
                len(clauses), len(entities), contract_type,
            )

            return {
                "raw_text": raw_text,
                "clauses": clauses,
                "entities": entities,
                "contract_type": contract_type,
                "status": ReviewStatus.IN_PROGRESS,
            }

        except Exception as e:
            logger.error("条款提取Agent异常: %s", str(e))
            return {
                "errors": state.errors + [f"条款提取失败: {str(e)}"],
                "status": ReviewStatus.IN_PROGRESS,
                "raw_text": state.raw_text or "",
                "clauses": self._fallback_extraction(state.raw_text or ""),
                "entities": [],
                "contract_type": "未识别",
            }

    def _prepare_text(self, raw_text: str, sections: list[dict]) -> str:
        if len(raw_text) > 15000:
            return raw_text[:15000] + "\n...(文本已截断)"
        return raw_text

    def _extract_with_llm(self, text: str) -> dict:
        messages = [
            SystemMessage(content=SYSTEM_PROMPT),
            HumanMessage(content=f"请提取以下合同的条款信息：\n\n{text}"),
        ]
        response = self.llm.invoke(messages)
        content = response.content.strip()

        if content.startswith("```"):
            content = content.split("\n", 1)[1]
            if content.endswith("```"):
                content = content[:-3]

        return json.loads(content)

    def _parse_clauses(self, result: dict) -> list[Clause]:
        clauses = []
        for item in result.get("clauses", []):
            try:
                category = ClauseCategory(item.get("category", "other"))
            except ValueError:
                category = ClauseCategory.OTHER
            clauses.append(
                Clause(
                    title=item.get("title", "未命名条款"),
                    content=item.get("content", ""),
                    category=category,
                    section_number=item.get("section_number", ""),
                )
            )
        return clauses

    def _parse_entities(self, result: dict) -> list[ContractEntity]:
        entities = []
        for item in result.get("entities", []):
            entities.append(
                ContractEntity(
                    entity_type=item.get("entity_type", "other"),
                    value=item.get("value", ""),
                    location=item.get("location", ""),
                )
            )
        return entities

    def _fallback_extraction(self, raw_text: str) -> list[Clause]:
        """LLM调用失败时的回退方案：基于规则切分条款。"""
        sections = parse_raw_text_to_sections(raw_text)
        return [
            Clause(
                title=s["title"],
                content=s["content"],
                category=ClauseCategory.OTHER,
            )
            for s in sections
        ]

File: golang/internal/agent/clause_extraction.go
package agent

import (
	"encoding/json"
	"fmt"
	"log"
	"regexp"
	"strings"

	"github.com/bcefghj/multi-agent-contract-review/golang/internal/config"
	"github.com/bcefghj/multi-agent-contract-review/golang/internal/model"
	"github.com/google/uuid"
)

// ClauseExtractionAgent 条款提取Agent — Go版
// 使用Eino或HTTP Client调用大模型API
type ClauseExtractionAgent struct {
	llmClient *config.LLMClient
}

func NewClauseExtractionAgent(client *config.LLMClient) *ClauseExtractionAgent {
	return &ClauseExtractionAgent{llmClient: client}
}

func (a *ClauseExtractionAgent) Name() string {
	return "ClauseExtractionAgent"
}

func (a *ClauseExtractionAgent) Process(state *model.ReviewState) error {
	log.Printf("[%s] 开始处理 reviewId=%s", a.Name(), state.ReviewID)

	if state.RawText == "" {
		state.Errors = append(state.Errors, "合同文本为空")
		return nil
	}

	text := state.RawText
	if len(text) > 15000 {
		text = text[:15000] + "\n...(已截断)"
	}

	systemPrompt := `你是专业的法律合同条款提取专家。请提取条款信息，输出JSON：
{"contract_type":"类型","clauses":[{"title":"","content":"","category":"","section_number":""}],"entities":[{"entity_type":"","value":"","location":""}]}
category可选: payment,liability,confidentiality,termination,intellectual_property,dispute_resolution,force_majeure,warranty,indemnification,governing_law,other`

	response, err := a.llmClient.Chat(systemPrompt, "请提取以下合同的条款信息：\n\n"+text)
	if err != nil {
		log.Printf("[%s] LLM调用失败: %v, 使用回退方案", a.Name(), err)
		state.Clauses = fallbackExtraction(state.RawText)
		state.ContractType = "未识别"
		state.Errors = append(state.Errors, fmt.Sprintf("条款提取LLM失败: %v", err))
		return nil
	}

	jsonStr := extractJSON(response)
	var result struct {
		ContractType string `json:"contract_type"`
		Clauses      []struct {
			Title         string `json:"title"`
			Content       string `json:"content"`
			Category      string `json:"category"`
			SectionNumber string `json:"section_number"`
		} `json:"clauses"`
		Entities []struct {
			EntityType string `json:"entity_type"`
			Value      string `json:"value"`
			Location   string `json:"location"`
		} `json:"entities"`
	}

	if err := json.Unmarshal([]byte(jsonStr), &result); err != nil {
		log.Printf("[%s] JSON解析失败: %v", a.Name(), err)
		state.Clauses = fallbackExtraction(state.RawText)
		state.ContractType = "未识别"
		return nil
	}

	state.ContractType = result.ContractType
	for _, c := range result.Clauses {
		state.Clauses = append(state.Clauses, model.Clause{
			ID:            uuid.New().String()[:8],
			Title:         c.Title,
			Content:       c.Content,
			Category:      model.ClauseCategory(c.Category),
			SectionNumber: c.SectionNumber,
		})
	}
	for _, e := range result.Entities {
		state.Entities = append(state.Entities, model.ContractEntity{
			EntityType: e.EntityType,
			Value:      e.Value,
			Location:   e.Location,
		})
	}

	state.Status = model.StatusInProgress
	log.Printf("[%s] 完成：%d个条款，%d个实体", a.Name(), len(state.Clauses), len(state.Entities))
	return nil
}

func extractJSON(text string) string {
	start := strings.Index(text, "{")
	end := strings.LastIndex(text, "}")
	if start >= 0 && end > start {
		return text[start : end+1]
	}
	return text
}

func fallbackExtraction(text string) []model.Clause {
	var clauses []model.Clause
	pattern := regexp.MustCompile(`(第[一二三四五六七八九十百]+条\s*.+)`)
	matches := pattern.FindAllStringIndex(text, -1)

	if len(matches) == 0 && text != "" {
		clauses = append(clauses, model.Clause{
			ID:       uuid.New().String()[:8],
			Title:    "全文",
			Content:  text,
			Category: model.CatOther,
		})
		return clauses
	}

	for i, match := range matches {
		title := text[match[0]:match[1]]
		var content string
		if i+1 < len(matches) {
			content = strings.TrimSpace(text[match[1]:matches[i+1][0]])
		} else {
			content = strings.TrimSpace(text[match[1]:])
		}
		clauses = append(clauses, model.Clause{
			ID:       uuid.New().String()[:8],
			Title:    strings.TrimSpace(title),
			Content:  content,
			Category: model.CatOther,
		})
	}
	return clauses
}
